### ISSUE TITLE
`Types.stdlib_version` returns wrong version in Julia v1.8.1

### ISSUE THREAD

**Issue by giordano** [JuliaLang/Pkg.jl#3203]
```console
% ./julia --startup-file=no -e '@show VERSION; using Pkg.Types; @show stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8")'
VERSION = v"1.8.0"
stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8") = v"6.2.1+2"
% julia --startup-file=no -e '@show VERSION; using Pkg.Types; @show stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8")'  
VERSION = v"1.8.1"
stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8") = v"6.2.1+1"
```
but for both 1.8.0 and 1.8.1 the version of `GMP_jll` (UUID `781609d7-10c4-51f6-84f2-b8444358ff6d`) should be `6.2.1+2`: https://github.com/JuliaLang/julia/blob/v1.8.0/stdlib/GMP_jll/Project.toml, https://github.com/JuliaLang/julia/blob/v1.8.1/stdlib/GMP_jll/Project.toml.

**Comment by giordano** [comment]
BTW, this is only with Julia v1.8.1, not v1.8.0 nor nightly.

**Comment by giordano** [comment]
In the `release-1.8` branch I see https://github.com/JuliaLang/Pkg.jl/blob/release-1.8/src/HistoricalStdlibs.jl#L866, which explains the problem: #3110 hasn't been backported.  But then I have no idea how this worked in Julia v1.8.0, wasn't Pkg cut out of the same branch?

**Comment by giordano** [comment]
According to https://github.com/JuliaLang/julia/blob/v1.8.0/stdlib/Pkg.version, Pkg in Julia v1.8.0 was based on https://github.com/JuliaLang/Pkg.jl/tree/63f4405d17e11decd2e5eb786cc491322b68c58c, and
https://github.com/JuliaLang/Pkg.jl/blob/63f4405d17e11decd2e5eb786cc491322b68c58c/src/HistoricalStdlibs.jl#L866 which makes me even more confused about how this worked at all in v1.8.0.

**Comment by DilumAluthge** [comment]
My guess is that if the version matches Base.VERSION, it bypasses the historical stdlib info and just checks the stdlib directly.

What do stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8") and stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8") give when you run them on Julia 1.8.0 and 1.8.1?

**Comment by giordano** [comment]
```console
% ./julia --startup-file=no -e '@show VERSION; using Pkg.Types; @show stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.0")'
VERSION = v"1.8.0"
stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.0") = v"6.2.1+2"
% julia --startup-file=no -e '@show VERSION; using Pkg.Types; @show stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.1")'  
VERSION = v"1.8.1"
stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.1") = v"6.2.1+2"
```

**Comment by DilumAluthge** [comment]
And if you pass `v"1.8.1"` on Julia 1.8.0, and `v"1.8.0"` on Julia 1.8.1?

**Comment by giordano** [comment]
> And if you pass `v"1.8.1"` on Julia 1.8.0

```console
% ./julia --startup-file=no -e '@show VERSION; using Pkg.Types; @show stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.1")'
VERSION = v"1.8.0"
stdlib_version(Base.UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), v"1.8.1") = v"6.2.1+1"
```
but in general I wouldn't expect this be necessarily correct, asking the future version of a package, you know...

> and v"1.8.0" on Julia 1.8.1?

That's what I've shown already in the first post?

**Comment by DilumAluthge** [comment]
> That's what I've shown already in the first post?

Whoops yes

**Comment by DilumAluthge** [comment]
Ideally we'd just do https://github.com/JuliaLang/Pkg.jl/pull/3096, and then this info won't be tied to a specific Julia version.

**Comment by giordano** [comment]
Addressed by #3096 + https://github.com/JuliaPackaging/HistoricalStdlibVersions.jl
